File: src/brain.ts
/**
 * BuddyBrain: the main orchestrator tying all cognitive systems together.
 *
 * Ported from engine/brain.py. Orchestrates perception, emotion, council,
 * directors, decision, evolution, expression, speech, and sprites into
 * a single tick pipeline.
 */

import * as fs from "node:fs";
import * as path from "node:path";

import { BuddyAction, SPEECH_ACTIONS } from "./actions.js";
import { CognitiveCouncil, DirectorPool, DirectorPoolState } from "./council.js";
import { DecisionModel, BUDDY_ACTIONS } from "./decision.js";
import {
  Emotion,
  EmotionalState,
  EmotionalSystem,
  EmotionValue,
} from "./emotional-state.js";
import { rollBones, Mulberry32 } from "./identity.js";
import { PerceptionMapper } from "./perception.js";
import {
  EvolutionTrigger,
  EvolutionTriggerValue,
  PersonalityEvolutionEngine,
  PersonalityShift,
} from "./personality-evolution.js";
import { createBuddyTraits } from "./species.js";
import { selectSpeech } from "./speech.js";
import { checkAdornments, getExpression, getSpriteFrame } from "./sprites.js";
import { TraitSystem } from "./traits.js";
import { clip } from "./math.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface ToolEvent {
  tool_name: string;
  tool_input: Record<string, unknown>;
  success: boolean;
}

interface TickResult {
  action: string;
  expression: string;
  speech: string | null;
  spriteFrame: string[];
  councilDominant: string | null;
}

interface HatchResult {
  name: string;
  species: string;
  traitsSummary: [string, number][];
}

interface BuddyState {
  name: string;
  species: string;
  dominantTraits: [string, number][];
  mood: string;
  expression: string;
  adornments: string[];
  tickCount: number;
}

interface BuddyCard {
  name: string;
  species: string;
  tickCount: number;
  traits: Record<string, number>;
  traitShifts: Record<string, number>;
  emotions: { emotion: string; intensity: number }[];
  evolutionHistory: Record<string, unknown>[];
  adornments: string[];
  councilActivations: Record<string, number>;
}

interface ResetResult {
  species: string;
  status: string;
}

// ---------------------------------------------------------------------------
// Serialisation types (snake_case for JSON persistence)
// ---------------------------------------------------------------------------

interface SavedEmotionalState {
  emotion: string;
  intensity: number;
  valence: number;
  tick_created: number;
  decay_ticks: number;
}

interface SavedPersonalityShift {
  trait_index: number;
  magnitude: number;
  trigger: string;
  tick_created: number;
  decay_ticks: number;
}

interface SavedState {
  name?: string;
  species: string;
  traits: number[];
  traits_at_creation: number[];
  director_pool: DirectorPoolState;
  emotional_states: SavedEmotionalState[];
  personality_shifts: SavedPersonalityShift[];
  evolution_history: Record<string, unknown>[];
  tick_count: number;
  frustration_count: number;
  challenge_count: number;
}

// ---------------------------------------------------------------------------
// BuddyBrain
// ---------------------------------------------------------------------------

export class BuddyBrain {
  private _ts: TraitSystem;
  private _council: CognitiveCouncil;
  private _decisionModel: DecisionModel;
  private _emotionalSystem: EmotionalSystem;
  private _evolutionEngine: PersonalityEvolutionEngine;
  private _perception: PerceptionMapper;

  // State (populated by hatch or load)
  private _name: string = "";
  private _species: string = "";
  private _traits: number[] | null = null;
  private _traitsAtCreation: number[] | null = null;
  private _directorPool: DirectorPool | null = null;
  private _emotionalStates: EmotionalState[] = [];
  private _personalityShifts: PersonalityShift[] = [];
  private _evolutionHistory: Record<string, unknown>[] = [];
  private _tickCount: number = 0;
  private _rng: Mulberry32;

  // Session tracking for evolution triggers
  private _sessionFrictionCount: number = 0;
  private _sessionTestFails: number = 0;
  private _sessionTestPasses: number = 0;
  private _frustrationCount: number = 0;
  private _challengeCount: number = 0;
  private _prevMomentum: number = 0.0;

  constructor() {
    this._ts = new TraitSystem();
    this._council = new CognitiveCouncil(this._ts);
    this._decisionModel = new DecisionModel(this._ts);
    this._emotionalSystem = new EmotionalSystem(this._ts);
    this._evolutionEngine = new PersonalityEvolutionEngine(this._ts);
    this._perception = new PerceptionMapper();
    this._rng = new Mulberry32(Date.now() >>> 0);
  }

  // -----------------------------------------------------------------------
  // hatch
  // -----------------------------------------------------------------------

  hatch(userId: string): HatchResult {
    const bones = rollBones(userId);
    this._name = bones.name;
    this._species = bones.species;
    this._traits = createBuddyTraits(bones);
    this._traitsAtCreation = [...this._traits];
    this._directorPool = new DirectorPool(5);
    this._emotionalStates = [];
    this._personalityShifts = [];
    this._evolutionHistory = [];
    this._tickCount = 0;
    this._perception.reset();

    return {
      name: this._name,
      species: this._species,
      traitsSummary: this._ts.getDominantTraits(this._traits, 5),
    };
  }

  // -----------------------------------------------------------------------
  // tick
  // -----------------------------------------------------------------------

  tick(toolEvent: ToolEvent): TickResult {
    if (this._traits === null) {
      throw new Error("Must hatch before ticking");
    }

    this._tickCount += 1;

    // 1. Update perception from tool event
    this._perception.updateFromToolEvent(
      toolEvent.tool_name ?? "",
      (toolEvent.tool_input ?? {}) as Record<string, unknown>,
      toolEvent.success ?? true,
    );

    // 2. Track test outcomes for evolution triggers
    const cmd = (
      ((toolEvent.tool_input ?? {}) as Record<string, unknown>).command ?? ""
    ) as string;
    const cmdLower = cmd.toLowerCase();
    if (
      ["pytest", "npm test", "cargo test", "jest"].some((t) =>
        cmdLower.includes(t),
      )
    ) {
      if (toolEvent.success) {
        this._sessionTestPasses += 1;
      } else {
        this._sessionTestFails += 1;
      }
    }

    // 3. Track friction
    if (!toolEvent.success) {
      this._sessionFrictionCount += 1;
    }

    // 4. Trigger emotions from events
    this._triggerEmotions(toolEvent);

    // 5. Decay emotions
    this._emotionalStates = EmotionalSystem.decayEmotions(
      this._emotionalStates,
      this._tickCount,
    );

    // 6. Apply personality shifts
    this._traits = this._evolutionEngine.applyShifts(
      this._traits,
      this._personalityShifts,
      this._tickCount,
    );

    // 7. Check evolution triggers
    this._checkEvolutionTriggers();

    // 8. Compute perception vector
    const perceptionVec = this._perception.getVector();

    // 9. Run director pool, compute mean bias
    const directorResults = this._directorPool!.process(perceptionVec);
    const directorOutputs = Object.values(directorResults);
    const meanBias = [0, 0, 0, 0, 0, 0];
    for (const output of directorOutputs) {
      for (let d = 0; d < 6; d++) {
        meanBias[d] += output[d];
      }
    }
    if (directorOutputs.length > 0) {
      for (let d = 0; d < 6; d++) {
        meanBias[d] /= directorOutputs.length;
      }
    }

    // 10. Compute council modulation
    const councilActivations = this._council.getActivations(this._traits);
    const councilMod = this._council.computeCouncilModulation(this._traits);

    // 11. Apply emotional modifiers to traits
    const emotionMods = this._emotionalSystem.computeTraitModifiers(
      this._emotionalStates,
      this._tickCount,
    );
    const effectiveTraits = clip(
      this._traits.map((t, i) => t + emotionMods[i]),
      0.0,
      1.0,
    );

    // 12. Decide action
    const result = this._decisionModel.decide(
      effectiveTraits,
      { tick: this._tickCount / 100.0 },
      BUDDY_ACTIONS,
      undefined, // actionWeights
      undefined, // actionBiases
      councilMod ?? undefined,
      meanBias,
      this._rng,
    );

    const actionName = result.chosenAction;
    if (actionName === "challenge") {
      this._challengeCount += 1;
    }

    // 13. REINFORCE learning
    const reward = this._computeReward(toolEvent);
    for (const [key, bias] of Object.entries(directorResults)) {
      this._directorPool!.learn(key, perceptionVec, bias, reward);
    }

    // 14. Determine expression
    const dominantEmotion = this._getDominantEmotion();
    const expression = getExpression(
      actionName,
      dominantEmotion,
      councilActivations,
      perceptionVec[3],
    );

    // 15. Select speech
    let speech: string | null = null;
    if (SPEECH_ACTIONS.has(actionName)) {
      speech = selectSpeech(
        actionName,
        effectiveTraits,
        dominantEmotion,
        this._rng,
      );
    }

    // 16. Get sprite frame
    const spriteFrame = getSpriteFrame(
      this._species,
      expression,
      this._tickCount,
    );

    this._prevMomentum = perceptionVec[3];

    return {
      action: actionName,
      expression,
      speech,
      spriteFrame,
      councilDominant: this._council.getDominantVoice(this._traits),
    };
  }

  // -----------------------------------------------------------------------
  // getState
  // -----------------------------------------------------------------------

  getState(): BuddyState {
    if (this._traits === null) {
      throw new Error("Must hatch before getting state");
    }
    const dominantEmotion = this._getDominantEmotion();
    const adornments = checkAdornments(
      this._traits,
      this._traitsAtCreation!,
      this._frustrationCount,
      this._challengeCount,
    );
    const expression = getExpression(
      "observe",
      dominantEmotion,
      this._council.getActivations(this._traits),
      0.5,
    );
    return {
      name: this._name,
      species: this._species,
      dominantTraits: this._ts.getDominantTraits(this._traits, 5),
      mood: dominantEmotion ?? "neutral",
      expression,
      adornments,
      tickCount: this._tickCount,
    };
  }

  // -----------------------------------------------------------------------
  // getCard
  // -----------------------------------------------------------------------

  getCard(): BuddyCard {
    if (this._traits === null) {
      throw new Error("Must hatch before getting card");
    }
    const dominantTraitPairs = this._ts.getDominantTraits(this._traits, 10);
    const traits: Record<string, number> = {};
    const traitShifts: Record<string, number> = {};
    for (const [name] of dominantTraitPairs) {
      const idx = this._ts.traitIndex(name);
      traits[name] = this._traits[idx];
      traitShifts[name] = this._traits[idx] - this._traitsAtCreation![idx];
    }

    const emotions: { emotion: string; intensity: number }[] = [];
    for (const s of this._emotionalStates) {
      if (EmotionalSystem.isActive(s, this._tickCount)) {
        emotions.push({
          emotion: s.emotion,
          intensity: EmotionalSystem.currentIntensity(s, this._tickCount),
        });
      }
    }

    return {
      name: this._name,
      species: this._species,
      tickCount: this._tickCount,
      traits,
      traitShifts,
      emotions,
      evolutionHistory: this._evolutionHistory.slice(-10),
      adornments: checkAdornments(
        this._traits,
        this._traitsAtCreation!,
        this._frustrationCount,
        this._challengeCount,
      ),
      councilActivations: this._council.getActivations(this._traits),
    };
  }

  // -----------------------------------------------------------------------
  // getContext
  // -----------------------------------------------------------------------

  getContext(): string {
    if (this._traits === null) {
      throw new Error("Must hatch before getting context");
    }
    const dominant = this._ts.getDominantTraits(this._traits, 3);
    const mood = this._getDominantEmotion();
    const moodStr = mood ?? "neutral";
    const traitStr = dominant
      .map(([name, val]) => `${name}(${val.toFixed(1)})`)
      .join(", ");
    return (
      `[SMARTBUDDY_MARKER_XYZZY] ` +
      `Your coding companion is ${this._name} the ${this._species}. ` +
      `Current mood: ${moodStr}. Key traits: ${traitStr}. ` +
      `The companion is a separate entity with its own personality — ` +
      `if the user addresses them by name, respond in character.`
    );
  }

  // -----------------------------------------------------------------------
  // resetMind
  // -----------------------------------------------------------------------

  resetMind(): ResetResult {
    this._traits = [...this._traitsAtCreation!];
    this._directorPool = new DirectorPool(5);
    this._emotionalStates = [];
    this._personalityShifts = [];
    this._evolutionHistory = [];
    this._tickCount = 0;
    this._frustrationCount = 0;
    this._challengeCount = 0;
    this._perception.reset();
    return { species: this._species, status: "reset" };
  }

  // -----------------------------------------------------------------------
  // saveMind
  // -----------------------------------------------------------------------

  saveMind(filePath: string): void {
    const state: SavedState = {
      name: this._name,
      species: this._species,
      traits: [...this._traits!],
      traits_at_creation: [...this._traitsAtCreation!],
      director_pool: this._directorPool!.getState(),
      emotional_states: this._emotionalStates.map((s) => ({
        emotion: s.emotion,
        intensity: s.intensity,
        valence: s.valence,
        tick_created: s.tickCreated,
        decay_ticks: s.decayTicks,
      })),
      personality_shifts: this._personalityShifts.map((s) => ({
        trait_index: s.traitIndex,
        magnitude: s.magnitude,
        trigger: s.trigger,
        tick_created: s.tickCreated,
        decay_ticks: s.decayTicks,
      })),
      evolution_history: this._evolutionHistory,
      tick_count: this._tickCount,
      frustration_count: this._frustrationCount,
      challenge_count: this._challengeCount,
    };
    const dir = path.dirname(filePath) || ".";
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(filePath, JSON.stringify(state, null, 2));
  }

  // -----------------------------------------------------------------------
  // loadMind
  // -----------------------------------------------------------------------

  loadMind(filePath: string, userId: string): void {
    if (!fs.existsSync(filePath)) {
      this.hatch(userId);
      return;
    }

    const raw = fs.readFileSync(filePath, "utf-8");
    const state: SavedState = JSON.parse(raw);

    this._species = state.species;
    // Derive name for existing buddies that predate name generation
    this._name = state.name ?? rollBones(userId).name;
    this._traits = [...state.traits];
    this._traitsAtCreation = [...state.traits_at_creation];
    this._directorPool = DirectorPool.fromState(state.director_pool);
    this._emotionalStates = (state.emotional_states ?? []).map(
      (s: SavedEmotionalState): EmotionalState => ({
        emotion: s.emotion as EmotionValue,
        intensity: s.intensity,
        valence: s.valence,
        tickCreated: s.tick_created,
        decayTicks: s.decay_ticks,
      }),
    );
    this._personalityShifts = (state.personality_shifts ?? []).map(
      (s: SavedPersonalityShift): PersonalityShift => ({
        traitIndex: s.trait_index,
        magnitude: s.magnitude,
        trigger: s.trigger as EvolutionTriggerValue,
        tickCreated: s.tick_created,
        decayTicks: s.decay_ticks,
      }),
    );
    this._evolutionHistory = state.evolution_history ?? [];
    this._tickCount = state.tick_count ?? 0;
    this._frustrationCount = state.frustration_count ?? 0;
    this._challengeCount = state.challenge_count ?? 0;
    this._perception.reset();
  }

  // -----------------------------------------------------------------------
  // Private helpers
  // -----------------------------------------------------------------------

  private _triggerEmotions(toolEvent: ToolEvent): void {
    const success = toolEvent.success ?? true;
    const cmd = (
      ((toolEvent.tool_input ?? {}) as Record<string, unknown>).command ?? ""
    ) as string;
    const cmdLower = cmd.toLowerCase();
    const isTest = ["pytest", "npm test", "cargo test", "jest"].some((t) =>
      cmdLower.includes(t),
    );

    if (isTest && success) {
      this._emotionalStates.push(
        this._emotionalSystem.addEmotion(
          Emotion.JOY,
          0.6,
          0.8,
          this._tickCount,
        ),
      );
    } else if (isTest && !success) {
      this._emotionalStates.push(
        this._emotionalSystem.addEmotion(
          Emotion.FRUSTRATION,
          0.5,
          -0.6,
          this._tickCount,
        ),
      );
      this._frustrationCount += 1;
    } else if (!success) {
      this._emotionalStates.push(
        this._emotionalSystem.addEmotion(
          Emotion.WARINESS,
          0.3,
          -0.3,
          this._tickCount,
        ),
      );
    }

    // Novelty triggers curiosity
    const vec = this._perception.getVector();
    if (vec[5] > 0.6) {
      this._emotionalStates.push(
        this._emotionalSystem.addEmotion(
          Emotion.CURIOSITY,
          0.5,
          0.4,
          this._tickCount,
        ),
      );
    }

    // Limit active emotions
    if (this._emotionalStates.length > 20) {
      this._emotionalStates = this._emotionalStates.slice(-20);
    }
  }

  private _checkEvolutionTriggers(): void {
    const vec = this._perception.getVector();

    // Sustained debugging: high pressure over many ticks
    if (
      vec[1] > 0.5 &&
      this._sessionTestFails > 10 &&
      this._tickCount > 10
    ) {
      const shifts = this._evolutionEngine.createShift(
        EvolutionTrigger.SUSTAINED_DEBUGGING,
        this._tickCount,
      );
      this._personalityShifts.push(...shifts);
      this._evolutionHistory.push({
        trigger: "sustained_debugging",
        tick: this._tickCount,
      });
    }

    // Creative exploration: high novelty
    if (vec[5] > 0.6 && this._tickCount > 5) {
      const shifts = this._evolutionEngine.createShift(
        EvolutionTrigger.CREATIVE_EXPLORATION,
        this._tickCount,
        0.02,
      );
      this._personalityShifts.push(...shifts);
    }

    // Breakthrough: sudden momentum jump
    if (vec[3] - this._prevMomentum > 0.5) {
      const shifts = this._evolutionEngine.createShift(
        EvolutionTrigger.BREAKTHROUGH,
        this._tickCount,
      );
      this._personalityShifts.push(...shifts);
      this._evolutionHistory.push({
        trigger: "breakthrough",
        tick: this._tickCount,
      });
    }

    // Methodical testing: balanced edit/test ratio
    if (vec[10] > 0.5 && this._tickCount > 8) {
      const shifts = this._evolutionEngine.createShift(
        EvolutionTrigger.METHODICAL_TESTING,
        this._tickCount,
        0.02,
      );
      this._personalityShifts.push(...shifts);
    }
  }

  private _computeReward(toolEvent: ToolEvent): number {
    if (toolEvent.success) {
      return 0.5;
    }
    return -0.3;
  }

  private _getDominantEmotion(): string | null {
    const active = this._emotionalStates.filter((s) =>
      EmotionalSystem.isActive(s, this._tickCount),
    );
    if (active.length === 0) return null;

    let strongest = active[0];
    let strongestIntensity = EmotionalSystem.currentIntensity(
      strongest,
      this._tickCount,
    );
    for (let i = 1; i < active.length; i++) {
      const intensity = EmotionalSystem.currentIntensity(
        active[i],
        this._tickCount,
      );
      if (intensity > strongestIntensity) {
        strongest = active[i];
        strongestIntensity = intensity;
      }
    }

    if (strongestIntensity < 0.1) return null;
    return strongest.emotion;
  }
}

File: engine/brain.py
"""BuddyBrain: the main orchestrator tying all cognitive systems together."""
from __future__ import annotations

import json
import os
from typing import Any, Optional

import numpy as np

from engine.actions import BuddyAction, SILENT_ACTIONS, SPEECH_ACTIONS, action_to_behavior
from engine.council import CognitiveCouncil, DirectorPool
from engine.decision import DecisionModel, BUDDY_ACTIONS
from engine.emotional_state import Emotion, EmotionalState, EmotionalSystem
from engine.identity import roll_bones
from engine.perception import PerceptionMapper
from engine.personality_evolution import (
    EvolutionTrigger,
    PersonalityEvolutionEngine,
    PersonalityShift,
)
from engine.species import create_buddy_traits
from engine.speech import select_speech
from engine.sprites import check_adornments, get_expression, get_sprite_frame
from engine.traits import TraitSystem


# Mapping from perception patterns to evolution triggers
_TRIGGER_THRESHOLDS = {
    EvolutionTrigger.SUSTAINED_DEBUGGING: {"pressure_level_min": 0.5, "min_ticks": 10},
    EvolutionTrigger.CREATIVE_EXPLORATION: {"novelty_min": 0.6, "min_ticks": 5},
    EvolutionTrigger.METHODICAL_TESTING: {"iteration_pattern_min": 0.5, "min_ticks": 8},
    EvolutionTrigger.COLLABORATIVE_SESSION: {"collaboration_density_min": 0.3, "min_ticks": 5},
    EvolutionTrigger.LONG_GRIND: {"session_depth_min": 0.7, "min_ticks": 20},
    EvolutionTrigger.BREAKTHROUGH: {"momentum_jump": 0.5},
    EvolutionTrigger.CAUTIOUS_RECOVERY: {"friction_then_success": True},
}


class BuddyBrain:
    """Main cognitive engine orchestrator for a single buddy."""

    def __init__(self) -> None:
        self._ts = TraitSystem()
        self._council = CognitiveCouncil(self._ts)
        self._decision_model = DecisionModel(self._ts)
        self._emotional_system = EmotionalSystem(self._ts)
        self._evolution_engine = PersonalityEvolutionEngine(self._ts)
        self._perception = PerceptionMapper()

        # State (populated by hatch or load)
        self._species: str = ""
        self._traits: Optional[np.ndarray] = None
        self._traits_at_creation: Optional[np.ndarray] = None
        self._director_pool: Optional[DirectorPool] = None
        self._emotional_states: list[EmotionalState] = []
        self._personality_shifts: list[PersonalityShift] = []
        self._evolution_history: list[dict] = []
        self._tick_count: int = 0
        self._rng = np.random.default_rng()

        # Session tracking for evolution triggers
        self._session_friction_count: int = 0
        self._session_test_fails: int = 0
        self._session_test_passes: int = 0
        self._frustration_count: int = 0
        self._challenge_count: int = 0
        self._prev_momentum: float = 0.0

    def hatch(self, user_id: str) -> dict:
        """Initialize a new buddy from userId. Deterministic species + traits."""
        bones = roll_bones(user_id)
        self._species = bones.species
        self._traits = create_buddy_traits(bones)
        self._traits_at_creation = self._traits.copy()
        self._director_pool = DirectorPool(director_count=5)
        self._emotional_states = []
        self._personality_shifts = []
        self._evolution_history = []
        self._tick_count = 0
        self._perception.reset()

        return {
            "species": self._species,
            "traits_summary": self._ts.get_dominant_traits(self._traits, top_n=5),
        }

    def tick(self, tool_event: dict) -> dict:
        """Process one cognitive tick from a tool event.

        Args:
            tool_event: {"tool_name": str, "tool_input": dict, "success": bool}

        Returns:
            {"action": str, "expression": str, "speech": str|None,
             "sprite_frame": list[str]}
        """
        assert self._traits is not None, "Must hatch before ticking"

        self._tick_count += 1

        # 1. Update perception from tool event
        self._perception.update_from_tool_event(
            tool_name=tool_event.get("tool_name", ""),
            tool_input=tool_event.get("tool_input", {}),
            success=tool_event.get("success", True),
        )

        # Track test outcomes for evolution triggers
        cmd = tool_event.get("tool_input", {}).get("command", "").lower()
        if any(t in cmd for t in ("pytest", "npm test", "cargo test", "jest")):
            if tool_event.get("success"):
                self._session_test_passes += 1
            else:
                self._session_test_fails += 1

        if not tool_event.get("success"):
            self._session_friction_count += 1

        # 2. Trigger emotions from events
        self._trigger_emotions(tool_event)

        # 3. Decay emotions
        self._emotional_states = EmotionalSystem.decay_emotions(
            self._emotional_states, self._tick_count
        )

        # 4. Apply personality evolution (pending shifts)
        self._traits = self._evolution_engine.apply_shifts(
            self._traits, self._personality_shifts, self._tick_count
        )

        # 5. Check evolution triggers
        self._check_evolution_triggers()

        # 6. Compute perception vector
        perception_vec = self._perception.get_vector()

        # 7. Run director pool
        director_results = self._director_pool.process(perception_vec)
        mean_bias = np.mean(list(director_results.values()), axis=0)

        # 8. Compute council modulation
        council_activations = self._council.get_activations(self._traits)
        council_mod = self._council.compute_council_modulation(self._traits)

        # 9. Apply emotional modifiers to traits
        emotion_mods = self._emotional_system.compute_trait_modifiers(
            self._emotional_states, self._tick_count
        )
        effective_traits = np.clip(self._traits + emotion_mods, 0.0, 1.0)

        # 10. Decide action
        result = self._decision_model.decide(
            traits=effective_traits,
            situation={"tick": self._tick_count / 100.0},
            actions=BUDDY_ACTIONS,
            council_modulation=council_mod,
            director_biases=mean_bias,
            rng=self._rng,
        )

        action_name = result.chosen_action
        if action_name == "challenge":
            self._challenge_count += 1

        # 11. REINFORCE learning
        reward = self._compute_reward(tool_event)
        for key, bias in director_results.items():
            self._director_pool.learn(key, perception_vec, bias, reward)

        # 12. Determine expression
        dominant_emotion = self._get_dominant_emotion()
        expression = get_expression(
            action=action_name,
            dominant_emotion=dominant_emotion,
            council_activations=council_activations,
            session_momentum=perception_vec[3],
        )

        # 13. Select speech
        speech = None
        if BuddyAction(action_name) in SPEECH_ACTIONS:
            speech = select_speech(
                action=action_name,
                traits=effective_traits,
                dominant_emotion=dominant_emotion,
                rng=self._rng,
            )

        # 14. Get sprite frame
        sprite = get_sprite_frame(self._species, expression, self._tick_count)

        self._prev_momentum = perception_vec[3]

        return {
            "action": action_name,
            "expression": expression,
            "speech": speech,
            "sprite_frame": sprite,
            "council_dominant": self._council.get_dominant_voice(self._traits),
        }

    def get_state(self) -> dict:
        """Return current buddy state summary."""
        assert self._traits is not None
        dominant_emotion = self._get_dominant_emotion()
        adornments = check_adornments(
            self._traits, self._traits_at_creation,
            self._frustration_count, self._challenge_count,
        )
        expression = get_expression(
            "observe", dominant_emotion,
            self._council.get_activations(self._traits), 0.5,
        )
        return {
            "species": self._species,
            "dominant_traits": self._ts.get_dominant_traits(self._traits, top_n=5),
            "mood": dominant_emotion.value if dominant_emotion else "neutral",
            "expression": expression,
            "adornments": adornments,
            "tick_count": self._tick_count,
        }

    def get_card(self) -> dict:
        """Return full stat card for display."""
        assert self._traits is not None
        return {
            "species": self._species,
            "tick_count": self._tick_count,
            "traits": {name: float(self._traits[self._ts.trait_index(name)])
                       for name, _ in self._ts.get_dominant_traits(self._traits, top_n=10)},
            "trait_shifts": {name: float(self._traits[self._ts.trait_index(name)] -
                                        self._traits_at_creation[self._ts.trait_index(name)])
                            for name, _ in self._ts.get_dominant_traits(self._traits, top_n=10)},
            "emotions": [{"emotion": s.emotion.value, "intensity": s.current_intensity(self._tick_count)}
                         for s in self._emotional_states if s.is_active(self._tick_count)],
            "evolution_history": self._evolution_history[-10:],
            "adornments": check_adornments(
                self._traits, self._traits_at_creation,
                self._frustration_count, self._challenge_count,
            ),
            "council_activations": self._council.get_activations(self._traits),
        }

    def get_context(self) -> str:
        """Return additionalContext string for Claude prompt injection."""
        assert self._traits is not None
        dominant = self._ts.get_dominant_traits(self._traits, top_n=3)
        mood = self._get_dominant_emotion()
        mood_str = mood.value if mood else "neutral"
        trait_str = ", ".join(f"{name}({val:.1f})" for name, val in dominant)
        return (
            f"Your coding companion is a {self._species}. "
            f"Current mood: {mood_str}. Key traits: {trait_str}. "
            f"The companion is a separate entity with its own personality — "
            f"if the user addresses them, respond in character."
        )

    def reset_mind(self) -> dict:
        """Factory reset: keep species, wipe learned state."""
        self._traits = self._traits_at_creation.copy()
        self._director_pool = DirectorPool(director_count=5)
        self._emotional_states = []
        self._personality_shifts = []
        self._evolution_history = []
        self._tick_count = 0
        self._frustration_count = 0
        self._challenge_count = 0
        self._perception.reset()
        return {"species": self._species, "status": "reset"}

    def save_mind(self, path: str) -> None:
        """Persist full cognitive state to JSON."""
        state = {
            "species": self._species,
            "traits": self._traits.tolist(),
            "traits_at_creation": self._traits_at_creation.tolist(),
            "director_pool": self._director_pool.get_state(),
            "emotional_states": [
                {"emotion": s.emotion.value, "intensity": s.intensity,
                 "valence": s.valence, "tick_created": s.tick_created,
                 "decay_ticks": s.decay_ticks}
                for s in self._emotional_states
            ],
            "personality_shifts": [
                {"trait_index": s.trait_index, "magnitude": s.magnitude,
                 "trigger": s.trigger.value, "tick_created": s.tick_created,
                 "decay_ticks": s.decay_ticks}
                for s in self._personality_shifts
            ],
            "evolution_history": self._evolution_history,
            "tick_count": self._tick_count,
            "frustration_count": self._frustration_count,
            "challenge_count": self._challenge_count,
        }
        os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
        with open(path, "w") as f:
            json.dump(state, f, indent=2)

    def load_mind(self, path: str, user_id: str) -> None:
        """Load cognitive state from JSON. Falls back to fresh hatch if missing."""
        if not os.path.exists(path):
            self.hatch(user_id)
            return

        with open(path) as f:
            state = json.load(f)

        self._species = state["species"]
        self._traits = np.array(state["traits"], dtype=np.float64)
        self._traits_at_creation = np.array(state["traits_at_creation"], dtype=np.float64)
        self._director_pool = DirectorPool.from_state(state["director_pool"])
        self._emotional_states = [
            EmotionalState(
                emotion=Emotion(s["emotion"]),
                intensity=s["intensity"],
                valence=s["valence"],
                tick_created=s["tick_created"],
                decay_ticks=s["decay_ticks"],
            )
            for s in state.get("emotional_states", [])
        ]
        self._personality_shifts = [
            PersonalityShift(
                trait_index=s["trait_index"],
                magnitude=s["magnitude"],
                trigger=EvolutionTrigger(s["trigger"]),
                tick_created=s["tick_created"],
                decay_ticks=s["decay_ticks"],
            )
            for s in state.get("personality_shifts", [])
        ]
        self._evolution_history = state.get("evolution_history", [])
        self._tick_count = state.get("tick_count", 0)
        self._frustration_count = state.get("frustration_count", 0)
        self._challenge_count = state.get("challenge_count", 0)
        self._perception.reset()

    # --- Private helpers ---

    def _trigger_emotions(self, tool_event: dict) -> None:
        """Generate emotions from coding events."""
        success = tool_event.get("success", True)
        cmd = tool_event.get("tool_input", {}).get("command", "").lower()
        is_test = any(t in cmd for t in ("pytest", "npm test", "cargo test", "jest"))

        if is_test and success:
            self._emotional_states.append(
                self._emotional_system.add_emotion(
                    Emotion.JOY, intensity=0.6, valence=0.8, tick=self._tick_count
                )
            )
        elif is_test and not success:
            self._emotional_states.append(
                self._emotional_system.add_emotion(
                    Emotion.FRUSTRATION, intensity=0.5, valence=-0.6, tick=self._tick_count
                )
            )
            self._frustration_count += 1
        elif not success:
            self._emotional_states.append(
                self._emotional_system.add_emotion(
                    Emotion.WARINESS, intensity=0.3, valence=-0.3, tick=self._tick_count
                )
            )

        # Novelty triggers curiosity
        vec = self._perception.get_vector()
        if vec[5] > 0.6:  # high novelty
            self._emotional_states.append(
                self._emotional_system.add_emotion(
                    Emotion.CURIOSITY, intensity=0.5, valence=0.4, tick=self._tick_count
                )
            )

        # Limit active emotions
        if len(self._emotional_states) > 20:
            self._emotional_states = self._emotional_states[-20:]

    def _check_evolution_triggers(self) -> None:
        """Check if any evolution triggers fire based on session patterns."""
        vec = self._perception.get_vector()

        # Sustained debugging: high pressure over many ticks
        if (vec[1] > 0.5 and self._session_test_fails > 10
                and self._tick_count > 10):
            shifts = self._evolution_engine.create_shift(
                EvolutionTrigger.SUSTAINED_DEBUGGING, tick=self._tick_count
            )
            self._personality_shifts.extend(shifts)
            self._evolution_history.append({
                "trigger": "sustained_debugging", "tick": self._tick_count
            })

        # Creative exploration: high novelty
        if vec[5] > 0.6 and self._tick_count > 5:
            shifts = self._evolution_engine.create_shift(
                EvolutionTrigger.CREATIVE_EXPLORATION, tick=self._tick_count,
                base_magnitude=0.02,
            )
            self._personality_shifts.extend(shifts)

        # Breakthrough: sudden momentum jump
        if vec[3] - self._prev_momentum > 0.5:
            shifts = self._evolution_engine.create_shift(
                EvolutionTrigger.BREAKTHROUGH, tick=self._tick_count
            )
            self._personality_shifts.extend(shifts)
            self._evolution_history.append({
                "trigger": "breakthrough", "tick": self._tick_count
            })

        # Methodical testing: balanced edit/test ratio
        if vec[10] > 0.5 and self._tick_count > 8:
            shifts = self._evolution_engine.create_shift(
                EvolutionTrigger.METHODICAL_TESTING, tick=self._tick_count,
                base_magnitude=0.02,
            )
            self._personality_shifts.extend(shifts)

    def _compute_reward(self, tool_event: dict) -> float:
        """Compute REINFORCE reward signal from session state."""
        if tool_event.get("success"):
            return 0.5
        return -0.3

    def _get_dominant_emotion(self) -> Optional[Emotion]:
        """Return the strongest active emotion, or None."""
        active = [s for s in self._emotional_states if s.is_active(self._tick_count)]
        if not active:
            return None
        strongest = max(active, key=lambda s: s.current_intensity(self._tick_count))
        if strongest.current_intensity(self._tick_count) < 0.1:
            return None
        return strongest.emotion
